test/helpers/value_hook.dart

```diff
@@ -1,42 +1,198 @@
 import 'dart:async';
 
-import 'package:test/test.dart' as test show expect, Matcher;
-
-import 'wait_for.dart';
+import 'package:test/test.dart' as test show expect, Matcher, TestFailure;
 
 /// Useful to easily test handler functions
 /// that set values outside of their scope.
 ///
 /// ```dart
-/// // Construct VHook with initial value of 'false'
-/// VHook<bool> calledHandler = VHook<bool>(false);
+/// // Construct VHook with no initial value
+/// VHook calledHandler = VHook();
 ///
 /// willCallHandler(() {
-///   calledHandler.set(true);
 ///   // Some handler code
+///   // ...
+///   calledHandler.complete();
 /// });
 ///
 /// // Test that the handler was called
-/// // This will throw an error if the handler was not called.
-/// calledHandler.expect(equals(true));
+/// // If the handler was not called after 5s, throw an error
+/// calledHandler.awaitCompletion(Duration(seconds: 5));
 /// ```
 class VHook<T> {
   /// The value of the variable.
-  T _value;
+  dynamic _value;
+
+  /// The error that might be completed with
+  Object _error = _VHookValue.none;
+
+  /// Flag that indicates that the value reached it's final state.
+  final _completer = Completer<T>();
+
+  /// Stream used to subscribe to value changes
+  ///
+  /// Used to allow awaiting values with [awaitValue]
+  final _stream = StreamController<T>.broadcast();
 
   /// Construct a new ValueHook with the given initial value.
-  VHook(this._value);
+  ///
+  /// If not initial value is specified, a placeholder will
+  /// be used. Please check [hasValue] before directly accessing
+  /// [value] in such a case.
+  VHook([this._value = _VHookValue.none]);
 
   /// Set the value of the variable.
-  void set(T val) => _value = val;
+  void set(T val) {
+    // Do not change value if already completed
+    _assertNotCompleted();
+
+    // Update value
+    _value = val;
+
+    // Notify listeners
+    _stream.add(val);
+  }
+
+  /// Complete with the given value.
+  void completeValue(T val) {
+    // Update value
+    set(val);
+
+    // Complete with newly updated value
+    complete();
+  }
+
+  /// Complete with the current value.
+  void complete() {
+    // Can not complete multiple times
+    _assertNotCompleted();
+
+    // Close stream
+    _stream.close();
+
+    // Complete value
+    _completer.complete(_value);
+  }
+
+  /// Complete with an error or exception
+  void completeError(Object error) {
+    // Can not complete multiple times
+    _assertNotCompleted();
+
+    // Save error
+    _error = error;
```

Comment:
Why do u save the error?

Reply:
For better user experience. You can always use call `VHook.error` to access the `Object` used in `hook.completeError(...)`.